aptos-move/block-executor/src/view.rs

```diff
@@ -115,58 +190,20 @@ impl<'a, T: Transaction, X: Executable> ParallelState<'a, T, X> {
-                            // one with lowest txn_idx. It observed a dependency, so some thread
-                            // aborted dep_idx. If that abort returned execution task, by
-                            // minimality (lower transactions aren't waiting), that thread would
-                            // finish execution unblock txn_idx, contradiction. Otherwise,
-                            // execution_idx in scheduler was lower at a time when at least the
-                            // thread that aborted dep_idx was alive, and again, since lower txns
-                            // than txn_idx are not blocked, so the execution of dep_idx will
-                            // eventually finish and lead to unblocking txn_idx, contradiction.
-                            let (lock, cvar) = &*dep_condition;
-                            let mut dep_resolved = lock.lock();
-                            while let DependencyStatus::Unresolved = *dep_resolved {
-                                dep_resolved = cvar.wait(dep_resolved).unwrap();
-                            }
-                            if let DependencyStatus::ExecutionHalted = *dep_resolved {
-                                return ReadResult::HaltSpeculativeExecution(
-                                    "Speculative error to halt BlockSTM early.".to_string(),
-                                );
-                            }
-                        },
-                        DependencyResult::ExecutionHalted => {
-                            return ReadResult::HaltSpeculativeExecution(
-                                "Speculative error to halt BlockSTM early.".to_string(),
-                            );
-                        },
-                        DependencyResult::Resolved => continue,
+                    if !self.wait_for_dependency(txn_idx, dep_idx) {
+                        break;
                     }
                 },
                 Err(DeltaApplicationFailure) => {
-                    // Delta application failure currently should never happen. Here, we assume it
-                    // happened because of speculation and return 0 to the Move-VM. Validation will
-                    // ensure the transaction re-executes if 0 wasn't the right number.
-
-                    self.captured_reads
-                        .borrow_mut()
-                        .push(ReadDescriptor::from_speculative_failure(key.clone()));
-
-                    return ReadResult::HaltSpeculativeExecution(
-                        "Delta application failure (must be speculative)".to_string(),
-                    );
+                    self.captured_reads.borrow_mut().mark_failure();
+                    break;
                 },
             };
         }
+
+        ReadResult::HaltSpeculativeExecution(
+            "Delta application failure (must be speculative)".to_string(),
```

Comment:
This is not just delta failure anymore, right?

Reply:
Indeed, it's much cleaner to just return when needed.

Reply:
aaand I was breaking after waiting for dependency, instead of continue.. 🤦